rfd/0155-scoped-webauthn-credentials.md

```diff
@@ -0,0 +1,197 @@
+intended.
+
+For example, webauthn credentials intended for per-session MFA could be used
+instead for login. This means that if an attacker gets a hold of a user's
+webauthn credentials before the client consumes them, an attacker with the
+user's password could login as the user. Adding scope to the webauthn
+credentials would prevent the attacker from using the stolen webauthn
+credentials freely.
+
+Adding scope also enables us to allow webauthn credentials to be reused within
+specific scopes. For example, for admin actions it may make sense to allow the
+client to reuse a single webauthn credential for a string of requests that are
+functionally a single action, such as creating a new user and their corresponding
+invite token.
+
+## Details
+
+### Security
+
+Enforcing scope for webauthn credentials strictly improves security, as it is a
+purely restrictive change that prevents certain attack vectors, as described
+in [Why](#why).
+
+On the other hand, allowing clients to reuse webauthn credentials could have
+negative security implications if not handled carefully. Specifically, the
+client and server must clearly define:
+
+1. The webauthn credential's scope - client provided
+2. The number of reuses permitted - client provided
+3. The expiration of the credentials - server enforced (5 minutes)
+
+Additionally, reusing credentials will be limited to specific scopes where it
+is required. For now, this is only "admin actions", so any other scope will be
+forbidden from reusing webauthn credentials by the Auth server. Sensitive operations
+like "login" and "recovery" must never allow reuse.
+
+### Proto
+
+```proto
+// webauthn.proto
+
+message SessionData {
+  bytes challenge = 1 [(gogoproto.jsontag) = "challenge,omitempty"];
+  bytes user_id = 2 [(gogoproto.jsontag) = "userId,omitempty"];
+  repeated bytes allow_credentials = 3 [(gogoproto.jsontag) = "allowCredentials,omitempty"];
+  bool resident_key = 4 [(gogoproto.jsontag) = "residentKey,omitempty"];
+  string user_verification = 5 [(gogoproto.jsontag) = "userVerification,omitempty"];
+
+  // new:
+  // Scope authorized by this webauthn session. Ex: "login".
+  Scope scope = 6 [(gogoproto.jsontag) = "scope,omitempty"];
+  // UsesRemaining is a count of how many more times this session can be used for authentication
+  // before the session expires and is deleted. When set to 0 (default) or 1, the session will
+  // be deleted after the first authentication.
+  int64 uses_remaining = 7;
+}
+
+// Scope is a scope authorized by a webauthn session.
+enum Scope {
+  SCOPE_UNSPECIFIED = 0;
+  SCOPE_LOGIN = 1;
+  SCOPE_PASSWORDLESS_LOGIN = 2;
+  // MFA device management
+  SCOPE_MANAGE_DEVICES = 3;
+  // Account recovery - reset password, devices, etc.
+  SCOPE_RECOVERY = 4;
+  // Used for per-session MFA and moderated session presence checks
+  SCOPE_SESSION = 5;
+  SCOPE_HEADLESS = 6;
+  SCOPE_ADMIN_ACTION = 7;
```

Comment:
SCOPE_ADMIN_ACTION is the more flexible scope. Should we include advice for future features that might be tempted to "abuse" it instead of defining a new scope?